Import graph output type from protobuf

daquexian opened the pull request:
The graph output shape/elem_type is discarded when translating from protobuf to mlir. It is harmful to round-trip.

I renamed and polished the function `ImportInputTensorType` and applied it to graph outputs.

chentong319 reviewed (approved):
LGTM

chentong319 commented:
@jenkins-droid test this please.

tjingrant commented on src/Builder/FrontendDialectTransformer.cpp:
```diff
@@ -61,20 +61,17 @@ class FrontendGenImpl {
     std::vector<int64_t> dims;
-    auto shape_proto = input.type().tensor_type().shape();
-    auto input_tensor_legalized_name = legalize_name(input.name());
+    auto shape_proto = value_info.type().tensor_type().shape();
```
I have similar concerns. I believe current output type is deduced via shape/type inference (as opposed to taken without checking from the onnx model specification). Even though the onnx model spec may record information about the output type, it seems like the type information we derived is going to be more accurate/reliable.

daquexian replied:
I have updated the code so that the type is imported only if the shape is presented in protobuf.

> Even though the onnx model spec may record information about the output type, it seems like the type information we derived is going to be more accurate/reliable.

I disagree. Shape inference is not always reliable. For example, current onnx-mlir cannot infer shapes for mobilenetv2.

tjingrant replied:
Hi @daquexian, my concern is that according to the ONNX python API documentation:
https://github.com/onnx/onnx/blob/master/docs/PythonAPIOverview.md#creating-an-onnx-model-using-helper-functions

You can put in anything for output shape and make an onnx model out of it. That's why I'm dubious about its correctness. Besides, your patch forces shape inference to happen on the operation whose output is returned, which effectively will always override the output type imported anyways, so why bother importing the output type?

daquexian replied:
@tjingrant In the often-encountered cases that shape inference is not applicable, the output shape from onnx model is still useful. I'm implementing onnx optimizer based on onnx-mlir, the shape inference is not necessary in my case (for example, shape inference may fail on some complicated models, but these models can still be optimized), but the output shape is always needed to translate the mlir back to protobuf and ensure no information is lost

daquexian commented:
@chentong319 @tjingrant 

I have fixed the backend test error. The reason is

1. onnx-mlir doesn't run shape inference on ops without dynamic shape output.
2. The shape and optional attributes are both inferred in shape inference pass.
3. Some passes use the value of optional attributes (which is null).

I think the shape inference and attribute inference should be done in two single passes? Shape inference is more complicated and not always doable (for example, a reshape op with a non-constant shape input). But the "attribute inference" is trivial. It just sets the optional attributes to the default value.

As a workaround, I force running the shape inference if the op is followed by a return op.

daquexian commented:
I fixed the CI error on onnx_lowering.mlir. It turns out the failed test case is taken from an incorrect example in onnx gather op docs (https://github.com/onnx/onnx/issues/2346, https://github.com/onnx/onnx/issues/2765, https://github.com/onnx/onnx/issues/1708). Since this PR enables shape inference on more ops, the inferred shape overwrites the original incorrect shape and then the check fails. I have updated the test to make the shape correct.

tjingrant commented:
> I think the shape inference and attribute inference should be done in two separate passes? Shape inference is more complicated and not always doable (for example, a reshape op with a non-constant shape input). But the "attribute inference" is trivial. It just sets the optional attributes to the default value.

@daquexian thanks a lot for your detailed investigation! This is a corner case that never crossed our minds, I will need to chat with other folks to tease out a more systematic fix. Thanks again for your work to uncover this issue.

tjingrant commented on src/Transform/ONNX/ShapeInferencePass.cpp:
```diff
@@ -34,7 +35,7 @@ class ShapeInferencePass
     // that return a dynamic shape.
     f.walk([&](mlir::Operation *op) {
-      if (returnsDynamicShape(op)) {
+      if (isUsedByReturnOp(op) || returnsDynamicShape(op)) {
```
May want to record why this is needed, it's not entirely obvious.

daquexian replied:
@tjingrant Thanks for your review! I added some comments here

tjingrant commented:
@jenkins-droid test this please.
